File: src/util/FormatUtils.php
<?php

namespace pocketcloud\cloud\util;

use Closure;

final class FormatUtils {

    public static function uptime(float $seconds): string {
        $days = 0;
        $hours = 0;
        $minutes = 0;

        while ($seconds >= 86400) {
            $days++;
            $seconds -= 86400;
        }

        while ($seconds >= 3600) {
            $hours++;
            $seconds -= 3600;
        }

        while ($seconds >= 60) {
            $minutes++;
            $seconds -= 60;
        }

        return ($days > 0 ? $days . "d, " : "") .
            ($hours > 0 ? $hours . "h, " : "") .
            ($minutes > 0 ? $minutes . "m, " : "") .
            ($seconds > 0 ? floor($seconds) . "s" : "");
    }

    public static function tps(float $tps, bool $suffix = true): string {
        $tps = round($tps, 2);
        $suffix = ($suffix ? " ticks/s" : "");
        if ($tps < 0) return "§c???";
        else if ($tps >= 17) return "§a" . $tps . $suffix;
        else if ($tps >= 12) return "§6" . $tps . $suffix;
        return "§c" . $tps . $suffix;
    }

    public static function bytes(int $bytes, ?int $maxBytes = null, ?float &$percent = null): string {
        if ($bytes < 0) return "§c???";

        $units = ["B", "KB", "MB", "GB", "TB", "PB", "EB"];
        $exp = $bytes > 0 ? (int) floor(log($bytes, 1024)) : 0;
        $value = $bytes / (1024 ** $exp);
        $value = round($value, 2);
        $formatted = $value . " " . $units[$exp];

        if ($maxBytes === null || $maxBytes <= 0) return $formatted;

        $percent = ($bytes / $maxBytes) * 100;
        if ($percent < 60) {
            $color = "§a";
        } else if ($percent < 85) {
            $color = "§e";
        } else {
            $color = "§c";
        }

        return sprintf("§b%s §8(%s%.1f%%§8)§r", $formatted, $color, $percent);
    }

    public static function usagePercentage(float $percentage, bool $higherBetter = false, int $precision = 3): string {
        if ($percentage < 0) return "§c???";

        $formatted = round($percentage, $precision) . "%";
        if ($percentage < 60) {
            $color = ($higherBetter ? "§c" : "§a");
        } else if ($percentage < 85) {
            $color = "§e";
        } else {
            $color = ($higherBetter ? "§a" : "§c");
        }

        return sprintf("%s%s", $color, $formatted);
    }

    public static function interpolate(string $subject, array $args, string $pattern = "{}"): string {
        foreach ($args as $arg) {
            $pos = strpos($subject, $pattern);
            if ($pos === false) break;
            $subject = substr_replace($subject, (string) $arg, $pos, strlen($pattern));
        }

        return $subject;
    }

    public static function seconds(float $s, int $precision = 3): string {
        if ($s >= 60) {
            return round($s / 60, $precision) . "min";
        } else if ($s >= 1) {
            return round($s, $precision) . "s";
        } else if ($s >= 0.001) {
            return round($s * 1000, $precision) . "ms";
        } else if ($s >= 0.000001) {
            return round($s * 1_000_000, $precision) . "µs";
        } else {
            return round($s * 1_000_000_000, $precision) . "ns";
        }
    }

    public static function milliseconds(float $ms, int $precision = 3): string {
        if ($ms >= 1000) {
            return round($ms / 1000, $precision) . "s";
        } else if ($ms >= 1) {
            return round($ms, $precision) . "ms";
        } else if ($ms >= 0.001) {
            return round($ms * 1000, $precision) . "µs";
        } else {
            return round($ms * 1_000_000, $precision) . "ns";
        }
    }

    /**
     * @param array $array
     * @param string $separator
     * @param string $keyValueSeparator
     * @param Closure(mixed $key): mixed|null $keyProcessHandler
     * @param Closure(mixed $processedKey, mixed $value): mixed|null $valueProcessHandler
     * @param string ...$ignoredKeys
     * @return string
     */
    public static function implodeWithKeys(array $array, string $separator = ", ", string $keyValueSeparator = ": ", ?Closure $keyProcessHandler = null, ?Closure $valueProcessHandler = null, string ...$ignoredKeys): string {
        $parts = [];

        foreach ($array as $key => $value) {
            if (in_array($key, $ignoredKeys)) continue;
            $actualKey = $key;
            if ($keyProcessHandler !== null) $key = $keyProcessHandler($key);
            if ($valueProcessHandler !== null) $value = $valueProcessHandler($actualKey, $value);
            $parts[] = $key . $keyValueSeparator . $value;
        }

        return implode($separator, $parts);
    }
}

File: cloud-common/src/main/java/de/pocketcloud/common/util/FormatUtils.java
package de.pocketcloud.common.util;

import java.util.regex.Matcher;

public final class FormatUtils {

    public static String uptime(double seconds) {
        int days = 0;
        int hours = 0;
        int minutes = 0;

        while (seconds >= 86400) {
            days++;
            seconds -= 86400;
        }

        while (seconds >= 3600) {
            hours++;
            seconds -= 3600;
        }

        while (seconds >= 60) {
            minutes++;
            seconds -= 60;
        }

        return (days > 0 ? days + "d, " : "") +
                (hours > 0 ? hours + "h, " : "") +
                (minutes > 0 ? minutes + "m, " : "") +
                (seconds > 0 ? (int) Math.floor(seconds) + "s" : "");
    }

    public static String tps(Double tps) {
        return tps(tps, true, true);
    }

    public static String tps(Double tps, boolean coloured) {
        return tps(tps, true, coloured);
    }

    public static String tps(Double tps, boolean suffix, boolean coloured) {
        if (tps == null || tps < 0) return (coloured ? "§c" : "") + "???";
        tps = round(tps, 2);

        String actualSuffix = suffix ? " ticks/s" : "";

        if (coloured) {
            if (tps >= 17) {
                return "§a" + tps + actualSuffix;
            } else if (tps >= 12) {
                return "§6" + tps + actualSuffix;
            }

            return "§c" + tps + actualSuffix;
        } else return tps + actualSuffix;
    }

    public static String bytes(Long bytes) {
        return bytes(bytes, null, false, true);
    }

    public static String bytes(Long bytes, boolean coloured) {
        return bytes(bytes, null, false, coloured);
    }

    public static String bytes(Long bytes, boolean higherPctBetter, boolean coloured) {
        return bytes(bytes, null, higherPctBetter, coloured);
    }

    public static String bytes(Long bytes, Long maxBytes, boolean higherPctBetter, boolean coloured) {
        if (bytes == null || bytes < 0) return (coloured ? "§c" : "") + "???";
        String[] units = {"B", "KB", "MB", "GB", "TB", "PB", "EB"};
        int exp = bytes > 0 ? (int) Math.floor(Math.log(bytes) / Math.log(1024)) : 0;
        double value = bytes / Math.pow(1024, exp);
        value = round(value, 2);
        String formatted = value + " " + units[exp];
        if (maxBytes == null || maxBytes <= 0) return formatted;
        double percent = ((double) bytes / maxBytes) * 100;

        String color = "";
        if (coloured) {
            if (percent < 60) {
                color = !higherPctBetter ? "§a" : "§c";
            } else if (percent < 85) {
                color = "§e";
            } else {
                color = higherPctBetter ? "§a" : "§c";
            }
        }

        String format = "%s (%s%.1f%%)";
        if (coloured) format = "§b%s §8(%s%.1f%%§8)§r";
        return String.format(format, formatted, color, percent);
    }

    public static String downloadSpeed(Double bytesPerSecond) {
        if (bytesPerSecond == null || Double.isNaN(bytesPerSecond) || Double.isInfinite(bytesPerSecond) || bytesPerSecond <= 0) return "N/A";
        if (bytesPerSecond >= Math.pow(1024, 3)) return String.format("%.1f GB/s", bytesPerSecond / Math.pow(1024, 3));
        if (bytesPerSecond >= Math.pow(1024, 2)) return String.format("%.1f MB/s", bytesPerSecond / Math.pow(1024, 2));
        if (bytesPerSecond >= 1024) return String.format("%.1f KB/s", bytesPerSecond / 1024);
        return String.format("%.0f B/s", bytesPerSecond);
    }

    public static String usagePercentage(Double percentage) {
        return usagePercentage(percentage, false, 3, true);
    }

    public static String usagePercentage(Double percentage, boolean coloured) {
        return usagePercentage(percentage, false, 3, coloured);
    }

    public static String usagePercentage(Double percentage, boolean higherBetter, boolean coloured) {
        return usagePercentage(percentage, higherBetter, 3, coloured);
    }

    public static String usagePercentage(Double percentage, boolean higherBetter, int precision, boolean coloured) {
        if (percentage == null || percentage < 0) return (coloured ? "§c" : "") + "???";
        String formatted = round(percentage, precision) + "%";
        String color = "";
        if (coloured) {
            if (percentage < 60) {
                color = higherBetter ? "§c" : "§a";
            } else if (percentage < 85) {
                color = "§e";
            } else {
                color = higherBetter ? "§a" : "§c";
            }
        }

        return color + formatted;
    }

    public static String interpolate(String subject, Object... args) {
        for (Object arg : args) subject = subject.replaceFirst("\\{}", Matcher.quoteReplacement(String.valueOf(arg)));
        return subject;
    }

    public static String seconds(double s) {
        return seconds(s, 3);
    }

    public static String seconds(double s, int precision) {
        if (s >= 60) {
            return round(s / 60, precision) + "min";
        } else if (s >= 1) {
            return round(s, precision) + "s";
        } else if (s >= 0.001) {
            return round(s * 1000, precision) + "ms";
        } else if (s >= 0.000001) {
            return round(s * 1_000_000, precision) + "µs";
        } else {
            return round(s * 1_000_000_000, precision) + "ns";
        }
    }

    public static String milliseconds(double ms) {
        return milliseconds(ms, 3);
    }

    public static String milliseconds(double ms, int precision) {
        if (ms >= 1000) {
            return round(ms / 1000, precision) + "s";
        } else if (ms >= 1) {
            return round(ms, precision) + "ms";
        } else if (ms >= 0.001) {
            return round(ms * 1000, precision) + "µs";
        } else {
            return round(ms * 1_000_000, precision) + "ns";
        }
    }

    private static double round(double value, int precision) {
        double scale = Math.pow(10, precision);
        return Math.round(value * scale) / scale;
    }
}